django/template/base.py

```diff
@@ -954,14 +954,10 @@ class NodeList(list):
     contains_nontext = False
 
     def render(self, context):
-        bits = []
-        for node in self:
-            if isinstance(node, Node):
-                bit = node.render_annotated(context)
-            else:
-                bit = node
-            bits.append(str(bit))
-        return mark_safe(''.join(bits))
+        return SafeString(''.join([
+            str(node.render_annotated(context))
```

Comment:
According to the `Node.render` docstring it should "Return the node rendered as a string" so if we really wanted to optimise for speed, we could potentially forgo the `str` as well, but I think probably better safe than sorry here, so better to leave it in.

Reply:
I'll leave the judgement call for how much to risk breaking to Carlton and Mariusz, but ... it's a great catch, and it does appear to be worth doing, if/when you can make the assumption that `render_annotated` always returns either `str` or `SafeString` (possibly defer it to another ticket at some point in the future if the risk is too high to stomach, understandably)

with `str()` call per item, in synthetic test:
```
%timeit render()
200 ms ± 2.29 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

without:
```
In [3]: %timeit render()
95.7 ms ± 201 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

I cannot for the life of me find the `str()` calls in cProfile output, so I dunno how to track it beyond timeit/walltime :)

Reply:
A minor touch of further due diligence and information; it looks like `ForNode` already assumes that everything coming back from a `render_annotated` call is something stringy, as it's implementation is:
```
nodelist = []
...
for node in self.nodelist_loop:
    nodelist.append(node.render_annotated(context))
...
return mark_safe(''.join(nodelist))
```
So it's notably missing the enforced `str()` call (and hey look Tim, it's our good friend `mark_safe`, put it on yer list if it's not already :) )

Reply:
Well, actually it'd better to enforce return from `Node.render()` string or list of strings. Multiple copies of strings leads to square complexity/